include/ctrace/ctrace.h

```diff
@@ -0,0 +1,141 @@
+#ifndef CTRACE_H
+#define CTRACE_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#if defined(__cplusplus)
+ #define CTRACE_BEGIN_DEFINITIONS extern "C" {
+ #define CTRACE_END_DEFINITIONS }
+#else
+ #define CTRACE_BEGIN_DEFINITIONS
+ #define CTRACE_END_DEFINITIONS
+#endif
+
+#ifdef _MSC_VER
+ #define CTRACE_FORCE_NO_INLINE __declspec(noinline)
+#else
+ #define CTRACE_FORCE_NO_INLINE __attribute__((noinline))
+#endif
+
+#ifdef _MSC_VER
+ #define CTRACE_FORCE_INLINE __forceinline
+#elif defined(__clang__) || defined(__GNUC__)
+ #define CTRACE_FORCE_INLINE __attribute__((always_inline)) inline
+#else
+ #define CTRACE_FORCE_INLINE inline
+#endif
+
+// See `CPPTRACE_PATH_MAX` for more info.
+#define CTRACE_PATH_MAX 4096
+
+// TODO: Add exports
+
+CTRACE_BEGIN_DEFINITIONS
+    typedef struct raw_trace ctrace_raw_trace;
+    typedef struct object_trace ctrace_object_trace;
+    typedef struct stacktrace ctrace_stacktrace;
+
+    // Represents a boolean value, ensures a consistent ABI.
+    typedef int8_t ctrace_bool;
+    // A type that can represent a pointer, alias for `uintptr_t`.
+    typedef uintptr_t ctrace_frame_ptr;
+    typedef struct object_frame ctrace_object_frame;
+    typedef struct stacktrace_frame ctrace_stacktrace_frame;
+    typedef struct safe_object_frame ctrace_safe_object_frame;
+
+    // Type-safe null-terminated string wrapper
+    typedef struct {
+        const char* data;
+    } ctrace_owning_string;
+
+    struct object_frame {
+        ctrace_frame_ptr raw_address;
+        ctrace_frame_ptr obj_address;
+        const char* obj_path;
```

Comment:
Looking at this with a fresh set of eyes, I think it's ok without even though it first struck me as inconsistent

Reply:
Yeah I didn't want the user to have to do ``frame.raw_address.data``, and since it's wrapped by another type, they wouldn't be freeing directly either way